File: 450 Questions for DSA/Love Babbar DSA/1. Array/12_Merge_Without_Extra_Space.cpp
//{ Driver Code Starts
#include <bits/stdc++.h> 
using namespace std; 


// } Driver Code Ends
class Solution{
    
    public:
        //Function to merge the arrays.
        
       int nextGap(int gap){
        if(gap <= 1)
            return 0;
        return (gap + 1) / 2;
    }

    void merge(long long arr1[], long long arr2[], int n, int m) 
    { 
        int gap = n + m;
        
        while(gap > 0){
            int i = 0;
            while(i + gap < n){
                if (arr1[i] > arr1[i + gap])
                    swap(arr1[i], arr1[i + gap]);
                i += 1;
            }
            
            int j = max(0, gap - n);
            
            while (i < n && j < m){
                if (arr1[i] > arr2[j])
                    swap(arr1[i], arr2[j]);
                i += 1;
                j += 1;
            }
            
            if (j < m){
                j = 0;
                while (j + gap < m){
                    if (arr2[j] > arr2[j + gap])
                        swap(arr2[j], arr2[j + gap]);
                    j += 1;
                }
            }
            gap = nextGap(gap);
        }
    }
};

//{ Driver Code Starts.

int main() 
{ 
	
	int T;
	cin >> T;
	
	while(T--){
	    int n, m;
	    cin >> n >> m;
	    
	    long long arr1[n], arr2[m];
	    
	    for(int i = 0;i<n;i++){
	        cin >> arr1[i];
	    }
	    
	    for(int i = 0;i<m;i++){
	        cin >> arr2[i];
	    }
	    Solution ob;
	    ob.merge(arr1, arr2, n, m); 

        for (int i = 0; i < n; i++) 
            cout<<arr1[i]<<" "; 
        
       
	    for (int i = 0; i < m; i++) 
		    cout<<arr2[i]<<" "; 
	    
	    cout<<endl;
	}

	return 0; 
} 

// } Driver Code Ends

File: 450 Questions for DSA/Love Babbar DSA/1. Array/12_Merge_Without_Extra_Space.py
# // O(n*m) Time, O(1) Space

# def merge(arr1, arr2, n, m):
#     for i in range(n):
#         if arr1[i]>=arr2[0]:
#             arr1[i], arr2[0] = arr2[0], arr1[i]
#             arr2.sort()
#     # arr1.sort()
#     # print(arr1)
#     # print(arr2)

# O(nlog n) Time , O(1) Space
import math
def nextGap(gap):
 
    if (gap <= 1):
        return 0
    return (gap // 2) + (gap % 2)
 
 
def merge(arr1, arr2, n, m):
 
    gap = n + m
    gap = nextGap(gap)
    while gap > 0:
 
        # comparing elements in
        # the first array.
        i = 0
        while i + gap < n:
            if (arr1[i] > arr1[i + gap]):
                arr1[i], arr1[i + gap] = arr1[i + gap], arr1[i]
 
            i += 1
 
        # comparing elements in both arrays.
        j = gap - n if gap > n else 0
        while i < n and j < m:
            if (arr1[i] > arr2[j]):
                arr1[i], arr2[j] = arr2[j], arr1[i]
 
            i += 1
            j += 1
 
        if (j < m):
 
            # comparing elements in the
            # second array.
            j = 0
            while j + gap < m:
                if (arr2[j] > arr2[j + gap]):
                    arr2[j], arr2[j + gap] = arr2[j + gap], arr2[j]
 
                j += 1
 
        gap = nextGap(gap)

arr1 = list(map(int,input().split()))
n = len(arr1)
arr2 = list(map(int,input().split()))
m = len(arr2)
merge(arr1,arr2,n,m)
print(arr1)
print(arr2)

